### PR TITLE
Register reading from mcu

### PR DESCRIPTION
**Pull request by tannermpeterson**

### CODE REVIEW

File: tests/test_mc_comm.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+):
+    # mock sleep to speed up the test
+    mocked_sleep = mocker.patch.object(mc_comm, "sleep", autospec=True)
+
+    mc_process = four_board_mc_comm_process[0]
+    simulator = mantarray_mc_simulator_no_beacon[4]
+
+    # need to mock read here to have better control over the reads going into McComm
+    test_read_values = [
+        SERIAL_COMM_MAGIC_WORD_BYTES[:4],
```
Comment (eli88fine) [inline tests/test_mc_comm.py:160]: why are you using `4`? is it arbitrary, or half the size of the word? a comment might be useful

File: tests/test_mc_comm.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+        bytes(0),
```
Comment (eli88fine) [inline tests/test_mc_comm.py:161]: why are there 4 copies of 0 bytes? a comment might be useful

File: tests/test_mc_comm.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        bytes(0),
+        bytes(0),
+        bytes(0),
+        SERIAL_COMM_MAGIC_WORD_BYTES[4:],
```
Comment (eli88fine) [inline tests/test_mc_comm.py:165]: why are you using `4`? is it arbitrary, or half the size of the word? is it `length of magic word minus the amount truncated 5 lines earlier`? a comment might be useful

File: tests/test_mc_comm.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    confirm_queue_is_eventually_of_size(board_queues[3][2], 1)
+    error_queue.put(expected_error)
+    confirm_queue_is_eventually_of_size(error_queue, 1)
+
+    actual = mc_process.hard_stop()
+    assert actual["fatal_error_reporter"] == [expected_error]
+
+    confirm_queue_is_eventually_empty(board_queues[1][0])
+    confirm_queue_is_eventually_empty(board_queues[2][0])
+    confirm_queue_is_eventually_empty(board_queues[3][0])
```
Comment (eli88fine) [inline tests/test_mc_comm.py:61]: why were only these 5 chosen? is there some significance, or arbitrary? a comment might be useful here

File: tests/test_mc_comm.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    confirm_queue_is_eventually_empty(board_queues[0][0])
+    confirm_queue_is_eventually_empty(board_queues[0][2])
+
+    assert actual["board_1"]["main_to_instrument_comm"] == [expected[1][0]]
+    assert actual["board_2"]["main_to_instrument_comm"] == [expected[2][0]]
+    assert actual["board_3"]["main_to_instrument_comm"] == [expected[3][0]]
```
Comment (eli88fine) [inline tests/test_mc_comm.py:67]: why were only these 6 assertions chosen? is there some significance, or arbitrary? a comment might be useful here

File: tests/test_mc_comm.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+
+def test_McCommunicationProcess_register_magic_word__registers_magic_word_in_serial_comm_from_board__when_first_packet_is_not_truncated(
+    four_board_mc_comm_process, mantarray_mc_simulator_no_beacon, mocker
+):
+    mc_process = four_board_mc_comm_process[0]
+    testing_queue = mantarray_mc_simulator_no_beacon[3]
+    simulator = mantarray_mc_simulator_no_beacon[4]
+
+    board_idx = 0
+    invoke_process_run_and_check_errors(mc_process)
```
Comment (eli88fine) [inline tests/test_mc_comm.py:127]: why is this being run prior to the connection being set? it's not clear to me

File: tests/test_mc_comm.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+        bytes(0),
+        bytes(0),
+        SERIAL_COMM_MAGIC_WORD_BYTES[4:],
+    ]
+    mocked_read = mocker.patch.object(
+        simulator, "read", autospec=True, side_effect=test_read_values
+    )
+
+    board_idx = 0
+    invoke_process_run_and_check_errors(mc_process)
```
Comment (eli88fine) [inline tests/test_mc_comm.py:172]: why is this being run prior to the connection being set? it's not clear to me

File: tests/test_mc_comm.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+        bytes(0),
+        bytes(0),
+        bytes(0),
+        bytes(0),
+        bytes(0),
+    ]
+    mocker.patch.object(simulator, "read", autospec=True, side_effect=test_read_values)
+
+    board_idx = 0
+    invoke_process_run_and_check_errors(mc_process)
```
Comment (eli88fine) [inline tests/test_mc_comm.py:224]: why is this being run prior to the connection being set? it's not clear to me

File: tests/test_mc_comm.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+
+    mc_process = four_board_mc_comm_process[0]
+    simulator = mantarray_mc_simulator_no_beacon[4]
+
+    # need to mock read here to have better control over the reads going into McComm
+    test_read_values = [
+        SERIAL_COMM_MAGIC_WORD_BYTES[:-1],
+        bytes(0),
+        bytes(0),
+        bytes(0),
```
Comment (eli88fine) [inline tests/test_mc_comm.py:217]: why are there 5x0bytes here? is it related to some constant? (sleeping for 5 seconds?)

**Review by eli88fine — COMMENTED** [review]
looks good overall. some thoughts/questions in the comments

File: tests/fixtures_mc_comm.py [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+from .fixtures import generate_board_and_error_queues
+
+
+@pytest.fixture(scope="function", name="four_board_mc_comm_process")
+def fixture_four_board_mc_comm_process():
+    board_queues, error_queue = generate_board_and_error_queues(num_boards=4)
+    p = McCommunicationProcess(board_queues, error_queue)
+    yield p, board_queues, error_queue
+    # clean up queues to avoid broken pipe errors
+    p.hard_stop()
```
[the remark this answers is not in the document]
Comment (tannermpeterson) [reply tests/fixtures_mc_comm.py:16]: Most of our other test fixtures for process also have the hard_stop called in the fixture itself, including those for both of our simulators, so I was following what had already been done. I'm not sure why that's the case, but it probably seemed like a good idea at some point

File: tests/fixtures_mc_simulator.py [hunk 11]
```diff
@@ -7,13 +8,54 @@
...
+            read_timeout_seconds,
+        )
+        self._sleep_after_write_seconds = sleep_after_write_seconds
+
+    def write(self, input_item):
+        super().write(input_item)
+        if self._sleep_after_write_seconds is not None:
+            time.sleep(self._sleep_after_write_seconds)
+
+    def start(self) -> None:
```
[the remark this answers is not in the document]
Comment (tannermpeterson) [reply tests/fixtures_mc_simulator.py:42]: Couldn't we solve that problem by using `*args` and `**kwargs` in the subclass's `__init__`?

File: tests/fixtures_mc_simulator.py [hunk 12]
```diff
@@ -25,14 +67,23 @@ def fixture_mantarray_mc_simulator():
     yield input_queue, output_queue, error_queue, testing_queue, simulator
 
 
-class MantarrayMCSimulatorNoBeacons(MantarrayMCSimulator):
+class MantarrayMCSimulatorNoBeacons(MantarrayMCSimulatorSleepAfterWrite):
     def _send_status_beacon(self, truncate=False) -> None:
         self._time_of_last_status_beacon_secs = time.perf_counter()
         self._output_queue.put_nowait(bytes(0))
 
+    def start(self) -> None:
```
[the remark this answers is not in the document]
Comment (tannermpeterson) [reply tests/fixtures_mc_simulator.py:75]: pylint was complaining for some reason about it being an abstract method

File: tests/fixtures_mc_comm.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (tannermpeterson) [reply tests/fixtures_mc_comm.py:16]: I agree though that would should move away from that though to speed up our tests

File: tests/fixtures_mc_simulator.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (eli88fine) [reply tests/fixtures_mc_simulator.py:35]: yep, I think that would work

File: tests/fixtures_mc_simulator.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (tannermpeterson) [reply tests/fixtures_mc_simulator.py:68]: warning (W0223, abstract-method, MantarrayMcSimulatorNoBeacons) Method 'start' is abstract in class 'MantarrayMcSimulatorSleepAfterWrite' but is not overridden

File: tests/test_mc_comm.py [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+    assert actual[1] is simulator
+    assert actual[2] is None
+    assert actual[3] is None
+
+
+def test_McCommunicationProcess_register_magic_word__registers_magic_word_in_serial_comm_from_board__when_first_packet_is_truncated_to_more_than_8_bytes(
+    four_board_mc_comm_process, mantarray_mc_simulator_no_beacon, mocker
+):
+    mc_process = four_board_mc_comm_process[0]
+    testing_queue = mantarray_mc_simulator_no_beacon[3]
```
[the remark this answers is not in the document]
Comment (tannermpeterson) [reply tests/test_mc_comm.py:97]: Would we want to make this dictionary immutable?

[the remark this answers is not in the document]
Comment (eli88fine) [reply tests/test_mc_comm.py:97]: if the fixture is `function` scoped, I don't see a particular benefit to making it immutable---nothing a test did to the dictionary could hurt the fixture (or its teardown procedure) or carry over to a different test. But I may be missing the thrust of your question

[the remark this answers is not in the document]
Comment (tannermpeterson) [reply tests/test_mc_comm.py:97]: That clears things up

File: src/mantarray_desktop_app/constants.py [hunk 14]
```diff
@@ -236,6 +236,9 @@
 SERIAL_COMM_MAGIC_WORD_BYTES = b"CURI BIO"
 SERIAL_COMM_TIMESTAMP_LENGTH_BYTES = 8
 SERIAL_COMM_CHECKSUM_LENGTH_BYTES = 4
+SERIAL_COMM_MAX_PACKET_LENGTH_BYTES = (
```
Comment (eli88fine) [inline src/mantarray_desktop_app/constants.py:239]: does this include the magic word? or is it specifically supposed to exclude the magic word in what this is defining as the "length"?

File: src/mantarray_desktop_app/mc_comm.py [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+    """
+
+    def __init__(
+        self,
+        # pylint: disable=duplicate-code
+        board_queues: Tuple[  # pylint-disable: duplicate-code
+            Tuple[
+                Queue[Dict[str, Any]],  # pylint: disable=unsubscriptable-object
+                Queue[  # pylint: disable=unsubscriptable-object
+                    Dict[str, Any]
```
Comment (eli88fine) [inline src/mantarray_desktop_app/mc_comm.py:38]: could this be replaced with `*args, **kwargs` also (like your suggestion for the simulator)?

**Review by eli88fine — APPROVED** [review]
looks good. question about packet size, and minor note/question for future

File: src/mantarray_desktop_app/constants.py (hunk 14 above)
Comment (tannermpeterson) [reply src/mantarray_desktop_app/constants.py:239]: This includes the magic word

File: src/mantarray_desktop_app/mc_comm.py (hunk 15 above)
Comment (tannermpeterson) [reply src/mantarray_desktop_app/mc_comm.py:38]: I suppose it could, I can test it out in a new branch

Comment (tannermpeterson) [reply src/mantarray_desktop_app/mc_comm.py:38]: Seems to work fine

**Pull request merged**